### PR TITLE
Add @ThriftHeader for per invocation headers

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: drift-client/src/main/java/io/airlift/drift/client/DriftMethodHandler.java [hunk 1]
```diff
@@ -62,6 +73,20 @@ public boolean isAsync()
 
     public ListenableFuture<Object> invoke(Optional<String> addressSelectionContext, Map<String, String> headers, List<Object> parameters)
     {
+        if (!headerParameters.isEmpty()) {
+            headers = new LinkedHashMap<>(headers);
+            for (Entry<Integer, ThriftHeaderParameter> entry : headerParameters.entrySet()) {
+                headers.put(entry.getValue().getName(), (String) parameters.get(entry.getKey()));
```
Comment (electrum) [inline drift-client/src/main/java/io/airlift/drift/client/DriftMethodHandler.java]: Why is `parameters` not `List<String>`?

File: drift-codec/src/main/java/io/airlift/drift/codec/metadata/ThriftHeaderParameter.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+public class ThriftHeaderParameter
+{
+    private final int index;
+    private final String name;
+
+    public ThriftHeaderParameter(int index, String name)
+    {
+        checkArgument(index >= 0, "index is negative");
+        requireNonNull(name, "name is null");
```
Comment (electrum) [inline drift-codec/src/main/java/io/airlift/drift/codec/metadata/ThriftHeaderParameter.java]: Move this inline

File: drift-codec/src/main/java/io/airlift/drift/codec/metadata/ThriftMethodMetadata.java [hunk 3]
```diff
@@ -76,26 +80,55 @@ public ThriftMethodMetadata(Method method, ThriftCatalog catalog)
...
+                checkArgument(
+                        parameterTypes[parameterIndex] == String.class,
+                        "ThriftMethod [%s] parameter %s annotated with @ThriftHeader must be a String",
+                        methodName(method),
+                        parameterIndex);
+
+                String headerName = thriftHeader.value();
+                checkArgument(
+                        !headerName.isEmpty(),
+                        "ThriftMethod [%s] parameter %s @ThriftHeader.name can not be empty",
```
Comment (electrum) [inline drift-codec/src/main/java/io/airlift/drift/codec/metadata/ThriftMethodMetadata.java]: "can not" -> "cannot"

File: drift-codec/src/main/java/io/airlift/drift/codec/metadata/ThriftMethodMetadata.java [hunk 4]
```diff
@@ -76,26 +80,55 @@ public ThriftMethodMetadata(Method method, ThriftCatalog catalog)
...
+
+            ThriftHeader thriftHeader = Stream.of(parameterAnnotations[parameterIndex])
+                    .filter(ThriftHeader.class::isInstance)
+                    .map(ThriftHeader.class::cast)
+                    .findAny()
+                    .orElse(null);
+            if (thriftHeader != null) {
+                checkArgument(
+                        thriftField == null,
+                        "ThriftMethod [%s] parameter %s can not be annotated with both @ThriftField and @ThriftHeader",
```
Comment (electrum) [inline drift-codec/src/main/java/io/airlift/drift/codec/metadata/ThriftMethodMetadata.java]: "can not" -> "cannot" (or "must not")

File: drift-client/src/main/java/io/airlift/drift/client/DriftMethodHandler.java (hunk 1 above)
Comment (dain) [reply drift-client/src/main/java/io/airlift/drift/client/DriftMethodHandler.java]: Parameters are the parameters to the method being invoked... the arguments.  In the case of a header it is required to be a `String` but for `@Thrift` parameters they can be anything

**Pull request merged**